### PR TITLE
First attempts to address sites

### PR DESCRIPTION
**Pull request by mhines-usgs**

### CODE REVIEW

File: getNearestSegment.R [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  for (i in 1:nrow(reachLayer)) {
+    reachLayer@lines[[i]]@ID <- as.character(reachLayer[i, ]@data[[1]])
+  }
+  
+  ## create a bbox around the reaches
+  reachBbox <- st_bbox(reachLayer)
+  reachPoly <- as(raster::extent(reachBbox[[1]],reachBbox[[3]],reachBbox[[2]],reachBbox[[4]]), "SpatialPolygons")
+  
+  ## project poly layer to wgs84
+  crs(reachPoly) <- "+proj=longlat +datum=WGS84"
```
Comment (wdwatkins) [inline getNearestSegment.R]: This is from the `raster` package, so need to either load it or use `raster::crs`

File: getNearestSegment.R [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  
+  ## get the actual reachId and the original site coords
+  for (i in 1:nrow(dist.df)) {
+    dist.df$reachId[i] <- reachLayer[, ]@lines[[dist.df$ID[i]]]@ID
+    dist.df$site.lat[i] <- init_coords$lat[i]
+    dist.df$site.lon[i] <- init_coords$lon[i]
+  }
+
+  ## get the monitoring Location Identifier from the site coords
+  for (i in 1:nrow(dist.df)) {
```
Comment (wdwatkins) [inline getNearestSegment.R]: I think you could replace this loop with a `dplyr::left_join`

**Comment by mhines-usgs** [comment]
@wdwatkins I tried to bring the fist few lines of [this article ](https://www.r-bloggers.com/how-to-go-parallel-in-r-basics-tips/)to attempt to parallelize things but when I run this on Windows I get a firewall error. I can try it on my linux VM.

**Comment by wdwatkins** [comment]
That's annoying, probably because R has to spawn new processes and then communicate with them.  I suppose we don't have privileges to allow that?   https://github.com/tobigithub/R-parallel/wiki/R-parallel-Setups

**Comment by mhines-usgs** [comment]
@wdwatkins I made a copy of the original file to mess with setting up the dist2Line to be run in parallel. This seems to work ok locally for me now, but I'm not really sure how to decide how large n should be, is there some rationale for deciding based on the original dataset's size? https://github.com/USGS-R/2wp-temp-observations/pull/8/files#diff-52a9a88f7c0b9ab0779804ad7a6a15d2R64

**Comment by wdwatkins** [comment]
I think you would want `n` to be equal to `no_cores` —  that is the number of chunks the dataset is split into when the operation is parallelized.  Unless you are running out of RAM processing all the data at once, using more chunks is just increasing the overhead of sending information to different processes.

**Comment by mhines-usgs** [comment]
The latest getNearestSegmentParallel.R takes ~20 hours on my computer (only has 2 cores available in my linux VM out of the 4 on the machine) to address 19,921 points (subset from the original 628,527 wqp sites using a polygon clip of the reaches) against the reach layer with 456 segments, which is just the delaware segment layer. One suggestion I was given from a colleague was to look at simplifying the reach layer first to see if that might speed up the processing. Not sure if you want me to try to include that or if you think this will work for now @limnoliver?

**Comment by jread-usgs** [comment]
@mhines-usgs was the code I shared a dead end? It wasn't running in parallel and was taking a lot less time. I wonder if it is worth looking back into. As always, am available for questions/help.

**Comment by mhines-usgs** [comment]
No, that process isn't a dead end. I can adapt it and post it for comparison.

**Comment by jread-usgs** [comment]
here is 62000 sites (norwest) against 2716897 reaches:

```
ptm <- proc.time()
> remake::make('norwest_reaches', remake_file = 'remake_sites.yml')
<  MAKE > norwest_reaches
[    OK ] proj_string
[    OK ] nwm_segments
[    OK ] nwm_rivers
[    OK ] norwest_sites
[ BUILD ] norwest_reaches               |  norwest_reaches <- subset_closest(nwm_rivers, norwest_sites)
   user  system elapsed 
257.812   0.268 258.114 
> proc.time() - ptm
   user  system elapsed 
317.199   2.321 319.560 
```

**Comment by wdwatkins** [comment]
If simplifying the reach layer is acceptable for this purpose, I agree that would probably speed things up.

**Comment by limnoliver** [comment]
Thanks for implementing Jordan's code, Megan. I'll take a look this afternoon or tomorrow!

**Comment by wdwatkins** [comment]
@jread-usgs  was your code using `knnLookup`?  I think we should at least compare the outputs of using `knnLookup` to `dist2Line` — I'm pretty sure using the quad tree and `knnLookup` is only finding the nearest reach _vertex_, not the nearest reach _line_.  Note that `searchTrees::createTree` only takes point and rectangle data in the `dataType` argument, so it isn't treating the reaches as lines, just a collection of points.  I imagine that still does a decent job, but there could be places  where the nearest reach vertex and nearest reach line are not part of the same reach, and the point to reach mapping would be off.

**Comment by mhines-usgs** [comment]
I can save the outputs and compare them for similarities/differences

**Comment by jread-usgs** [comment]
yes, the code I shared is using `knnLookup`. Regarding the differences - I set mine up to find the closest reach centroid (centroid was the center of the reach's bounding box). This is a mapping for visual purposes only, and is a task that is distinct from "hydrologic addressing of sites for modeling purposes" which is a bigger lift and will require refinement. I hope that was clear in the definition of this task, let me know if it wasn't. Either way, considering the differences in speed, I wonder if `knnLookup` would still be faster if it were operating on each vertex (as opposed to each centroid), given it seemed the other method was several orders of magnitude slower.

**Comment by mhines-usgs** [comment]
Here is a csv of the results from each of the knnLookup as well as the dist2Line, I also added a neutral 3rd party of ArcMap's spatial join to settle ties. Kidding, but I was curious how they would differ, so I included those results.
https://drive.google.com/open?id=14yob33LgfUZCBKpBSCUpYUYpQ1IbzNUF

**Pull request merged**